File: kavach-backend/app/models/incident.py
import uuid
from datetime import datetime
from sqlalchemy import DateTime, Float, Text
from sqlalchemy.dialects.postgresql import ENUM, UUID as PgUUID
from sqlalchemy.orm import Mapped, mapped_column
from app.core.db import Base
from app.models.base import TimestampUUIDMixin

incident_status_enum = ENUM(
    "open", "graduated_1", "graduated_2", "graduated_3", "graduated_4",
    "resolved", "false_positive",
    name="incident_status", create_type=False,
)


class Incident(TimestampUUIDMixin, Base):
    __tablename__ = "incidents"
    elder_id: Mapped[uuid.UUID] = mapped_column(PgUUID(as_uuid=True), nullable=False)
    status: Mapped[str] = mapped_column(incident_status_enum, nullable=False, server_default="open")
    risk_score: Mapped[float] = mapped_column(Float, nullable=False, server_default="0")
    started_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False)
    resolved_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    resolution_note: Mapped[str | None] = mapped_column(Text, nullable=True)

File: kavach-android/app/src/main/java/com/kavach/app/domain/model/Incident.kt
package com.kavach.app.domain.model

/**
 * Domain model for an elder-safety incident.
 * Maps from both [IncidentEntity] (local) and future API responses.
 */
data class Incident(
    val id: String,
    val elderId: String,
    val status: IncidentStatus,
    val riskScore: Double,
    val startedAt: String,
    val resolvedAt: String? = null,
    val resolutionNote: String? = null,
)

enum class IncidentStatus(val value: String) {
    OPEN("open"),
    GRADUATED_1("graduated_1"),
    GRADUATED_2("graduated_2"),
    GRADUATED_3("graduated_3"),
    GRADUATED_4("graduated_4"),
    RESOLVED("resolved"),
    FALSE_POSITIVE("false_positive");

    val isActive: Boolean get() = this in listOf(OPEN, GRADUATED_1, GRADUATED_2, GRADUATED_3, GRADUATED_4)

    companion object {
        fun fromString(value: String): IncidentStatus =
            entries.firstOrNull { it.value == value } ?: OPEN
    }
}

data class ResolvedIncident(
    val incidentId: String,
    val status: String,
    val resolvedAt: String,
)

data class EvidencePackage(
    val incidentId: String,
    val pdfRef: String,
    val downloadUrl: String?,
)
